### ISSUE TITLE
Illegal repeated keyword in 01-pihole.conf

### ISSUE THREAD

**Issue by billyjthorton** [pi-hole/pi-hole#1521]
When installing pihole, I ran across an error that I can't seem to solve (or has been already solved).

When it gets to the end of the install script, the dnsmasq service cannot start (and coincidentally pihole). From the output of systemctl status dnsmasq.conf:

illegal repeated keyword at line 39 of .../01-pihole.conf

When looking at 01-pihole.conf, we get the error at the line:

cache-size=10000

There doesn't seem to really be any repeated keywords, so I'm not sure what the error really is. Is this a system error or a code error? Thanks for your help!

**Comment by dschaper** [comment]
That's an error from `dnsmasq` directly, either you have a pre-existing `cache-size` line in `/etc/dnsmasq.conf` or there is another file with that same directive. You can check the files, or run `pihole -d` and follow the on screen prompts to get us more info to debug with.

**Comment by billyjthorton** [comment]
In my `/etc/dnsmasq.conf` file I did not see any potential cache-size conflicts. Should I go ahead and send the debug info forward?

**Comment by dschaper** [comment]
Yes please, it should list out the configuration files, and we can see if we can find where the conflict is.

**Comment by billyjthorton** [comment]
:::  ---= Detecting Installed Package Versions:
:::       Pi-hole: v3.0.1 (master:v3.0.1-0-g26fcb1b)
:::       Pi-hole Dashboard: v3.0.1 (master:v3.0.1-0-ge204cea)
:::       lighttpd/1.4.35
:::       PHP 5.6.30-0+deb8u1 (cli) (built: Feb  8 2017 10:31:58)
:::
:::  ---= Detecting existence of /etc/pihole/setupVars.conf:
:::       File exists, parsing...done and was successfully sourced
:::
:::  ---= Detecting installed OS Distribution
:::
:::  ---= Checking processor variety
:::
:::  ---= Checking IPv6 Stack
:::       IPv6 addresses located on eth0
:::          No gateway detected.
:::
:::  ---= Checking IPv4 Stack
:::       IPv4 addresses located on eth0
:::        Pinging default IPv4 gateway: Gateway did not respond.
:::
:::  ---= Daemon Process Information
:::     Checking http port for lighttpd listener.
:::       Correctly configured.
:::
:::  ---= Daemon Process Information
:::     Checking domain port for dnsmasq listener.
:::       Correctly configured.
:::
:::  ---= Daemon Process Information
:::     Checking 4711 port for pihole-FTL listener.
:::  Failure: IPv4 Port not in use
:::
:::  ---= Processes Check
:::     Logging status of lighttpd, dnsmasq and pihole-FTL...
:::
:::  ---= Resolver Functions Check (IPv4)
shuf: /etc/pihole/gravity.list: No such file or directory
:::     Checking for necessary lighttpd files.
:::
:::  ---= Detecting existence of /etc/lighttpd/lighttpd.conf:
:::       File exists, parsing...done
:::
:::  ---= Detecting existence of /var/log/lighttpd/error.log:
:::       File exists, parsing...done
:::
:::
:::  ---= Detecting existence of /etc/dnsmasq.conf:
:::       File exists, parsing...done
:::
:::  ---= Detecting contents of /etc/dnsmasq.d/*:
:::
:::  ---= File /etc/dnsmasq.d/01-pihole.conf found
:::       Parsing...done
:::
:::  ---= File /etc/dnsmasq.d/README found
:::       Parsing...done
:::
:::  ---= File /etc/dnsmasq.d/SoftAp0 found
:::       Parsing...done
:::
:::
:::  ---= Detecting existence of /etc/pihole/whitelist.txt:
:::  /etc/pihole/whitelist.txt not found!
:::
:::  ---= Detecting existence of /etc/pihole/blacklist.txt:
:::  /etc/pihole/blacklist.txt not found!
:::
:::  ---= Detecting existence of /etc/pihole/adlists.list:
:::  /etc/pihole/adlists.list not found!
:::
:::  ---= Analyzing gravity.list
grep: /etc/pihole/gravity.list: No such file or directory
:::  Warning: No gravity.list file found!
:::
:::  ---= Analyzing pihole.log
:::  Warning: No pihole.log file found!
:::
:::  ---= Analyzing pihole-FTL.log
grep: /var/log/pihole-FTL.log: No such file or directory
:::  Warning: No pihole-FTL.log file found!
du: cannot access ‘/var/log/pihole-FTL.log’: No such file or directory
:::  Warning: No pihole-FTL.log file found!
tail: cannot open ‘/var/log/pihole-FTL.log’ for reading: No such file or directory
:::
::: --= User Action Required =--
::: Try loading a site that you are having trouble with now from a client web browser..
:::     (Press CTRL+C to finish logging.)
:::
:::  ---= pihole.log
::: Logging will automatically teminate in 60 seconds
^C:     50 seconds left.
::: Finishing debug write from interrupt... Quitting!
::: Finshed debugging!

**Comment by dschaper** [comment]
We need the token that was printed at the end of the run in order to look up your log upload.

**Comment by dschaper** [comment]
But from that I'm guessing you have an entry in `/etc/dnsmasq.d/SoftAp0`

**Comment by billyjthorton** [comment]
I apologize. The code is xauqd9o3v2

**Comment by dschaper** [comment]
`File /etc/dnsmasq.d/SoftAp0` has a duplicate cache size entry. Also your system is running `conmand` for connection management and that may interfere with `dnsmasq` if the DNS server is not disabled in `conman`.

**Comment by billyjthorton** [comment]
Awesome! I'll fix that right now. How do you disable the server in conman?

**Comment by dschaper** [comment]
Offhand I don't remember the process, but lets check and see what's running on DNS port 53. 
Can you run:

`sudo netstat -tulpn | grep 53` 

and well see what daemon is listening at that port.

**Comment by billyjthorton** [comment]
My second question and I've been banging my head for a bit now on it - I can ping www.google.com (or 8.8.8.8) as well as my default gateway, but when I run pihole -d, the gateway doesn't respond and I can't load any internet pages. Would you have any idea on how to fix this?

**Comment by dschaper** [comment]
it looks like you have some rather complex routing going on, the SoftAp0 is set for a few subnets, so I don't know if we'd be able to debug that. Maybe a look at `ip -4 route` and `ip -4 address`, `ip -4 link` and see if everything is as you expect?

**Comment by billyjthorton** [comment]
bringing this back up (sorry it's been a little while..) I did several fresh installs. Each time I have to change the SoftAp0 on a brand new linux install. Would it be easy enough to add a quick bash script to cancel out that repeated cache-size value?

Also I still am having trouble getting any web pages to resolve.
Upload Id: ryieix69qg
from pihole -d everything looks swell, but I don't understand why the pages don't resolve. Any insight you could provide would be extremely helpful! Thanks so much for the awesomeness in advanced!

**Comment by WaLLy3K** [comment]
Apologies for the delay in reply! :disappointed:

`sed '/cache-size/d' /etc/dnsmasq.d/SoftAp0` should be simple enough to delete the line containing that entry.

If you're still having issues with getting web pages to resolve:

1. I applaud you for sticking with Pi-hole despite what must be a frustrating issue
2. Could you please upload another debug log, providing us with a token please?

Thanks!

**Comment by kbarter** [comment]
This is the work around that I used:
[https://discourse.pi-hole.net/t/pi-holes-default-cache-directive-conflicts-with-cache-size-directive-in-softap0/8369](https://discourse.pi-hole.net/t/pi-holes-default-cache-directive-conflicts-with-cache-size-directive-in-softap0/8369)

> I know this is a late response, but I was having the same issue today. When Pi-Hole is installed, it’s files contain a cache-size directive. This conflicts with the cache-size directive in SoftAP0 that gets generated at boot time
> 
> I wanted to keep the functionality of providing DNS for the USB connections, so I updated the script /opt/scripts/boot/am335x_evm.sh by commenting out the following line:
> 
>  #echo "cache-size=2048" >> ${wfile}

**Comment by pralor** [comment]
This issue has been mentioned on **Pi-hole Userspace**. There might be relevant details there:

https://discourse.pi-hole.net/t/beaglebone-dns-will-not-start-after-update/27562/8

**Comment by oranges13** [comment]
Apologies for bringing back a dead thread, but I am having this same issue today on a Beagleboard Black and no matter what I modify, change, or comment out, the /etc/dnsmasq.d/SoftAp0 file *keeps being restored* with the erroneous cache-size line

I can manually remove it, restart all the services, but if there is a power failure or I reboot manually, the system fails to come back online because the file keeps getting restored.

I have modified the lines in /opt/scripts/boot/am335x_evm.sh to REMOVE the cache-size line addition but it's still being added to the file somehow.

I have tried making the file only writeable by root, but Pihole fails to start with a permissions error when it can't read that file.

**Comment by WarhawkCZ** [comment]
> Apologies for bringing back a dead thread, but I am having this same issue today on a Beagleboard Black and no matter what I modify, change, or comment out, the /etc/dnsmasq.d/SoftAp0 file _keeps being restored_ with the erroneous cache-size line
> 
> I can manually remove it, restart all the services, but if there is a power failure or I reboot manually, the system fails to come back online because the file keeps getting restored.
> 
> I have modified the lines in /opt/scripts/boot/am335x_evm.sh to REMOVE the cache-size line addition but it's still being added to the file somehow.
> 
> I have tried making the file only writeable by root, but Pihole fails to start with a permissions error when it can't read that file.

You're not alone. Google brought me here too. Any update?

**Comment by oranges13** [comment]
@WarhawkCZ I was able to change USE_GENERATED_DNSMAQ to no in /etc/default/bb-wl18xx.sh and that finally stopped overwriting any changes I made on reboot. I was using a slightly older image from March and that worked for me.
